adaptors_task/student/src/impl.cpp

```diff
@@ -1,8 +1,37 @@
 #include "impl.h"
 #include <vector>
-
+#include <stack>
+#include <iostream>
 const std::vector<std::pair<char, char>> brackets {
     {'(', ')'},
     {'[', ']'},
     {'{', '}'}
 };
+
+bool isValid(const std::string& source){
+    std::stack<char> notClosed;
+
+    for(auto i = 0; i < source.size(); i++){
+        bool closed = false;
+
+        if(!notClosed.empty()){
+            for(auto b : brackets){
+                if(std::make_pair(notClosed.top(), source[i]) == b){
+                    notClosed.pop();
+                    closed = true;
+                    break;
+                }  
+            }
+        }
+        
+        if(!closed){
+            for( auto b : brackets){
+                if(source[i] == b.first || source[i] == b.second){
```

Comment:
why should we check for closed brackets here, if we're in "non closed" branch?

Reply:
This was to check if the current character is a bracket and not some other character. It will not be added to the stack unless it's a bracket. Logically yes, I wouldn't need to add a closed bracket to the stack since it would mean it is an invalid string, but my invalid condition is whether the stack is empty or not at the end.